neatnet/nodes.py

```diff
@@ -466,9 +467,35 @@ def consolidate_nodes(
             return gdf
 
     # get clusters of nodes which should be consolidated
-    linkage = hierarchy.linkage(nodes.get_coordinates(), method="average")
-    nodes["lab"] = hierarchy.fcluster(linkage, tolerance, criterion="distance")
-    unique, counts = np.unique(nodes["lab"], return_counts=True)
+    indices = nodes.sindex.query(
+        nodes.geometry, predicate="dwithin", distance=tolerance
+    )
+    matrix = sparse.coo_array(
+        (np.ones(len(indices[0]), dtype=np.bool_), indices[::-1]),
+        shape=(len(nodes), len(nodes)),
+        dtype=np.bool_,
+    )
+    dwithin_graph = graph.Graph.from_sparse(matrix, ids=nodes.index)
+    comp_labels = dwithin_graph.component_labels
+    cardinalities = dwithin_graph.cardinalities
+
+    mask = cardinalities > 1
+    components = comp_labels[mask]
+    nodes_to_merge = nodes[mask]
```

Comment:
This could be also written as

```py
db = DBSCAN(eps=tolerance, min_samples=2).fit(nodes.get_coordinates())
comp_labels = db.labels_
mask = comp_labels > -1
components = comp_labels[mask]
nodes_to_merge = nodes[mask]
```

which is a bit simpler and takes roughly the same memory and time. But we do not directly depend on sklearn at the moment, though I believe it is a second order dependency anyway. Shall we use that for the sake of maintainability?

Reply:
Sounds good to me.

Reply:
@martin will you update that chunk in this PR or later?

Reply:
Just did